core/src/layers/prometheus.rs

```diff
@@ -98,784 +95,257 @@ use crate::*;
-
-        match res {
-            Ok(bytes) => {
-                self.stats
-                    .bytes_total
-                    .with_label_values(&labels)
-                    .observe(bytes.remaining() as f64);
-                Ok(bytes)
-            }
-            Err(e) => {
-                self.stats.increment_errors_total(self.op, e.kind());
-                Err(e)
-            }
-        }
+        self.operation_errors_total.with_label_values(&labels).inc();
     }
 }
 
-impl<R: oio::BlockingRead> oio::BlockingRead for PrometheusMetricWrapper<R> {
-    fn read(&mut self) -> Result<Buffer> {
-        let labels = self.stats.generate_metric_label(
-            self.scheme.into_static(),
-            Operation::BlockingReaderRead.into_static(),
-            &self.path,
-        );
-
-        let timer = self
-            .stats
-            .requests_duration_seconds
-            .with_label_values(&labels)
-            .start_timer();
-        let res = self.inner.read();
-        timer.observe_duration();
-
-        match res {
-            Ok(bs) => {
-                self.stats
-                    .bytes_total
-                    .with_label_values(&labels)
-                    .observe(bs.remaining() as f64);
-                Ok(bs)
-            }
-            Err(err) => {
-                self.stats.increment_errors_total(self.op, err.kind());
-                Err(err)
-            }
-        }
-    }
+struct OperationLabels<'a> {
+    scheme: Scheme,
+    namespace: &'a str,
+    root: &'a str,
+    op: Operation,
+    error: Option<ErrorKind>,
+    path: &'a str,
 }
 
-impl<R: oio::Write> oio::Write for PrometheusMetricWrapper<R> {
-    async fn write(&mut self, bs: Buffer) -> Result<()> {
-        let size = bs.len();
+impl<'a> OperationLabels<'a> {
+    fn names(error: bool, path_label_level: usize) -> Vec<&'a str> {
+        let mut names = Vec::with_capacity(6);
 
-        let labels = self.stats.generate_metric_label(
-            self.scheme.into_static(),
-            Operation::WriterWrite.into_static(),
-            &self.path,
-        );
+        names.extend([
+            observe::LABEL_SCHEME,
+            observe::LABEL_NAMESPACE,
+            observe::LABEL_ROOT,
+            observe::LABEL_OPERATION,
+        ]);
 
-        let timer = self
-            .stats
-            .requests_duration_seconds
-            .with_label_values(&labels)
-            .start_timer();
-        let res = self.inner.write(bs).await;
-        timer.observe_duration();
-
-        match res {
-            Ok(_) => {
-                self.stats
-                    .bytes_total
-                    .with_label_values(&labels)
-                    .observe(size as f64);
-                Ok(())
-            }
-            Err(err) => {
-                self.stats.increment_errors_total(self.op, err.kind());
-                Err(err)
-            }
+        if error {
+            names.push(observe::LABEL_ERROR);
```

Comment:
Is the order significant? Should we place `path` before `error`?

Reply:
The order of `path` and `error` is not important, what is important is that the order of label names and label values ​​should correspond, otherwise there will be problems